### PR TITLE
Added Standard VAE and Multinomial VAE models

### PR DESCRIPTION
**Pull request by kmussalim**
### Description
Added two notebooks with Standard VAE and Multinomial VAE to recommenders/examples/02_model_collaborative_filtering/

standard_vae_deep_dive
multinomial_vae_deep_dive
Also, added corresponding utils for this model to recommenders/reco_utils/recommender/vae:

multinomial_vae - contains classes for the model
standard_vae - contains classes for the model
sparse_vae - contains a function for binarization and a class for obtaining click matrix. This class is a modified version of AffinityMatrix class (can be found at recommenders/reco_utils/dataset/sparse.py /). There are 3 additional lines of code, which can be found by searching a comment "# LSECHANGE".
In this notebook we use Variational Autoencoders for recommending Top k items on MovieLens-1M dataset.
Also, the model has extension: using annealing to find optimal Beta. Using the model with optimal Beta has better performance than constant Beta = 1.

### Checklist:

- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [x] I have updated the documentation accordingly.
- [x] This PR is being made to `staging` and not `master`.

### CODE REVIEW

File: examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb [hunk 1]
```diff
@@ -0,0 +1,1829 @@
+{
```
Comment (miguelgfierro) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Pandas version: 1.0.5 
Tensorflow version: 2.2.0-rc1 
Keras version: 2.3.1 
</pre><p>what would be the difference if you used the library versions of the repo?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (miguelgfierro) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>please change the weight path to a temporary directory:</p><pre class="ql-syntax" spellcheck="false">from tempfile import TemporaryDirectory

tmp_dir = TemporaryDirectory()
WEIGHTS_PATH = os.path.join(tmp_dir, "mvae_weights.hdf5")
</pre><p> </p><p>what is the meaning of the comments:</p><pre class="ql-syntax" spellcheck="false"># DEFAULT_SEED - Overfits
# 98765 - Overfits
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (miguelgfierro) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>it is not clear if the code is something you developed or if it is based on other people work. If you develop the code, then it is ok to say "we use, we did, etc", otherwise, it is better to say "the authors use, the authors did..."</p><p> </p><p>Also, please remove all reference to "it is straightforward..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (miguelgfierro) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>&gt;&gt;&gt;&nbsp;To compute the gradients the reparametrization trick [2]&nbsp;</p><p> </p><p>What is [2]?</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (miguelgfierro) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>I would remove here the reference to ELBO, they are not actually using ELBO. They say it is an interpretation of ELBO, but it actually is a interpretation of the VAE loss function controlling the variational inference term. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (miguelgfierro) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>One thing that is missing in my view is a discussion on the key contributions and differences of Multi-VAE: the multinomial likelihood and the regularization using bayesian inference.</p><p> </p><p>They use multinomial likelihood which is good for modeling implicit feedback, and behaves better than logistic or gaussian likelihoods, which are typically more popular in the literature. By using multinomial likelihood, they are treating the problem as multi-class classification. As the authors discuss in the paper: "this likelihood rewards the model for putting probability mass on the non-zero entries in xu . But the model has a limited budget of probability mass, since π(zu) must sum to 1; the items must compete for this limited budget. The model should therefore assign more probability mass to items that are more likely to be clicked"</p><p> </p><p>They use bayesian inference for parameter estimation. The variational inference model q(zu|xu) approximates the intractable posterior distribution p(zu|xu). The authors interpret the standard VAE objective as a regularization parameter to better optimize the model.&nbsp;</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

**Comment by miguelgfierro** [comment]
guys this is really good job, I'm impressed. I did a first pass today and will continue over the next days.

@anargyri FYI

File: examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb (hunk 1 above)
Comment (kmussalim) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: We just realised that the model the notebook can be run with Pandas version: 0.25.3 and Tensorflow version: 1.15.2. So no difference on that.

We could not find a version of Keras used in the repo.

Comment (anargyri) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: I concur. There are many "we" whereas the text is following closely the papers. It would be better to say "according to" and cite Liang et al. or Higgins et al. as appropriate.

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>Could you link to a free version of the paper e.g. on arxiv. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>Missing curly bracket in softmax.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>The summation formula did not render properly when opening the notebook on GitHub.</p><p>"rerularization"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>Please state which ratings are 1and which 0.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>This section is taken almost verbatim from the paper. Please summarize in your own words and link to the relevant section in the paper for further reference.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>Could you refactor the users splitting code into a function defined in <code>python_splitters</code> ("get items from training set" could be an option).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>Remove the commented line.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (anargyri) [inline examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: <p>historicity -&gt; history</p><p> </p><p>It is a bit unclear where NDCG etc. are computed according to the paper (only the _te items are ranked, right?). Is this what <code>map_back_sparse()</code> doing or does it include the _tr items too? It would be good to clarify the protocol in the text.&nbsp;</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1191/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

**Comment by anargyri** [comment]
Could you also add unit tests for the utilities you have added?

File: examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb (hunk 1 above)
Comment (EvgeniaChroni) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: The model is developed from scratch based on the examined paper. 

We will make the necessary changes in order to be clear that we refer to the paper.

Comment (EvgeniaChroni) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: when we use different SEED the models may overfit because this causes different training/validation/test sets.

Comment (EvgeniaChroni) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: yes, it is possible.

Comment (EvgeniaChroni) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: _te dataset contains ratings and _tr are click matrices.

Comment (EvgeniaChroni) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: We should delete [2], is wrong

Comment (anargyri) [reply examples/02_model_collaborative_filtering/multi_vae_deep_dive.ipynb:1]: You should not get overfitting as long as the training, test , validation sets are separate (regardless of the value of the seed).

**Comment by miguelgfierro** [comment]
already in the repo https://github.com/microsoft/recommenders/pull/1194
